compiler/packages/babel-plugin-react-compiler/src/Entrypoint/Program.ts

```diff
@@ -430,11 +426,65 @@ export function compileProgram(
         prunedMemoValues: compiledFn.prunedMemoValues,
       });
     } catch (err) {
+      /**
+       * If an opt out directive is present, log only instead of throwing and don't mark as
+       * containing a critical error.
+       */
+      if (fn.node.body.type === 'BlockStatement') {
+        const optOutDirectives = findDirectiveDisablingMemoization(
+          fn.node.body.directives,
+        );
+        if (optOutDirectives.length > 0) {
+          logError(err, pass, fn.node.loc ?? null);
+          return null;
+        }
+      }
       hasCriticalError ||= isCriticalError(err);
       handleError(err, pass, fn.node.loc ?? null);
       return null;
     }
 
+    if (fn.node.body.type === 'BlockStatement') {
+      const optInDirectives = findDirectiveEnablingMemoization(
+        fn.node.body.directives,
+      );
+      const optOutDirectives = findDirectiveDisablingMemoization(
+        fn.node.body.directives,
+      );
```

Comment:
we could probably just calculate these once at the top of processFn, then reference the result here and above in the error case?